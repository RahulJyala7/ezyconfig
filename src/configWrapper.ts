import {ConfigValue, ResolvedConfig} from "./types/config";
import {createObjectPath, isPrimitive, isProxyWhitelist} from "./utils";

function wrap(path: string, config: ResolvedConfig): ResolvedConfig {
    return new Proxy(config, {
        set(target: ResolvedConfig, p: PropertyKey): boolean {
            if (Array.isArray(config)) {
                throw new Error(`Cannot change array ${path}. Config is immutable.`);
            } else {
                throw new Error(`Cannot change property ${createObjectPath(path, p.toString())}. Config is immutable.`);
            }
        },
        get(target: ResolvedConfig, p: PropertyKey, receiver: unknown): unknown|ConfigValue|ConfigValue[] {
            if (p === "toJSON") {
                return () => config;
            }

            if (Reflect.has(target, p) || isProxyWhitelist(p)) {
                return Reflect.get(target, p, receiver);
            }

            if (Array.isArray(config)) {
                throw new Error(
                    `Cannot get property ${path}[${p.toString()}] because it does not exist in the built config.`
                );
            } else {
                throw new Error(
                    `Cannot get property ${createObjectPath(path, p.toString())} because it does not exist in the built config.`
                );
            }
        }
    });
}

export function wrapInProxiesRecursive(path: string, config: ResolvedConfig): ResolvedConfig {
    if (Array.isArray(config)) {
        return wrap(path, config);
    }

    return wrap(path, Object.entries(config).reduce((acc, [key, value]) => ({
        ...acc,
        [key]: isPrimitive(value)
            ? value
            : wrapInProxiesRecursive(createObjectPath(path, key), value as ResolvedConfig)
    }), {}));
}
