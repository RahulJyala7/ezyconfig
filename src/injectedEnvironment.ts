import {CoreResolver} from "./resolvers/core";
import {ConfigValue, PublicCoreResolverInterface, Validator} from "./types/config";
import validators from "./validators";

export class InjectedEnvironment {
    public secret(environmentKey: string, defaultValue?: ConfigValue): PublicCoreResolverInterface {
        return new CoreResolver(environmentKey, defaultValue, true);
    }

    public value(environmentKey: string, defaultValue?: ConfigValue|ConfigValue[]): PublicCoreResolverInterface {
        return new CoreResolver(environmentKey, defaultValue, false);
    }

    public get validators(): {[key: string]: {name: string, fn: Validator}} {
        return validators;
    }
}
