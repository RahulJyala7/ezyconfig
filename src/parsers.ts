export function parseBoolean(value: string, isSecret: boolean): boolean|string {
    switch (value.toUpperCase().trim()) {
    case "TRUE":
    case "1":
    case "T": return true;
    case "FALSE":
    case "0":
    case "F": return false;
    default:
        return isSecret
            ? "Unable to parse secret value into boolean"
            : `Unable to parse value "${value}" into boolean`;
    }
}

export function parseNumber(value: string, isSecret: boolean): number|string {
    const num = parseInt(value, 10);
    if (isNaN(num)) {
        return isSecret
            ? "Unable to parse secret value into number"
            : `Unable to parse value "${value}" into number`;
    }
    return num;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function parseObject(value: string, isSecret: boolean): any|string {
    try {
        return JSON.parse(value);
    } catch (e) {
        return isSecret
            ? "Unable to parse secret value into JSON"
            : `Unable to parse value "${value}" into JSON`;
    }
}
