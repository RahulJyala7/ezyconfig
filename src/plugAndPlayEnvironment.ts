import {CoreResolver} from "./resolvers/core";
import {
    ConfigValueType,
    PlugAndPlayEnvironment,
    PublicArrayResolverInterface,
    PublicCoreResolverInterface
} from "./types/config";
import {Environment, EnvironmentArrayProperty, EnvironmentProperty} from "./types/environment";

type ResolverObject = {
    [key: string]: PublicCoreResolverInterface|PublicArrayResolverInterface
};

function isArrayType(resolver: EnvironmentProperty|EnvironmentArrayProperty): resolver is EnvironmentArrayProperty {
    return typeof (resolver as EnvironmentArrayProperty).splitOn !== "undefined";
}

export class PlugAndPlayEnvironmentImpl {
    private propertiesUsed: (PublicCoreResolverInterface|PublicArrayResolverInterface)[] = [];
    private environment: Environment;
    private resolvers: ResolverObject = {};

    constructor(env: Environment) {
        this.environment = env;
        this.buildResolvers();
    }

    private buildResolvers(): void {
        this.resolvers = this.environment.properties.reduce((acc, prop) => {
            if (acc[prop.name]) {
                throw new Error(
                    `${this.environment.name} PnP environment already has a definition for "${prop.name}"`
                );
            }

            let resolver: PublicCoreResolverInterface|PublicArrayResolverInterface =
                new CoreResolver(prop.envKey, prop.default, prop.secret);

            if (isArrayType(prop)) {
                resolver = resolver.asArray(prop.splitOn);
                switch (prop.type) {
                case ConfigValueType.BOOLEAN: resolver.ofBooleans(); break;
                case ConfigValueType.NUMBER: resolver.ofNumbers(); break;
                case ConfigValueType.OBJECT: resolver.ofObjects(); break;
                }
            } else {
                switch (prop.type) {
                case ConfigValueType.BOOLEAN: resolver.asBoolean(); break;
                case ConfigValueType.NUMBER: resolver.asNumber(); break;
                case ConfigValueType.OBJECT: resolver.asObject(); break;
                }
            }

            if (prop.validator) {
                resolver.validate(prop.validator);
            }

            return {
                ...acc,
                [prop.name]: resolver
            };
        }, {} as ResolverObject);
    }

    getValue(property: string): PublicCoreResolverInterface|PublicArrayResolverInterface {
        const resolver = this.resolvers[property];
        if (!resolver) {
            throw new Error(
                `${this.environment.name} PnP environment does not have "${property}" defined.`
            );
        }

        this.propertiesUsed.push(resolver);
        return resolver;
    }

    getPropertiesUsed(): (PublicCoreResolverInterface|PublicArrayResolverInterface)[] {
        return this.propertiesUsed;
    }
}

export function createPlugAndPlayEnvironment(env: Environment): {
    env: PlugAndPlayEnvironment, impl: PlugAndPlayEnvironmentImpl
} {
    const impl = new PlugAndPlayEnvironmentImpl(env);
    const envProxy = new Proxy({} as PlugAndPlayEnvironment, {
        get(target: PlugAndPlayEnvironment, p: PropertyKey, receiver: unknown): unknown {
            if (Reflect.has(target, p)) {
                return Reflect.get(target, p, receiver);
            }
            return impl.getValue(p.toString());
        },
        set(): boolean {
            throw new Error(`The PnP environment, ${env.name}, is immutable.`);
        }
    });

    return {impl, env: envProxy};
}
