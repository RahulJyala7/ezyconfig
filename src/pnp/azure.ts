import {ConfigValueType} from "../types/config";
import {Environment} from "../types/environment";

export = {
    name: "azure",
    properties: [
        {
            name: "account",
            envKey: "AZURE_ACCOUNT",
            secret: false,
            type: ConfigValueType.STRING
        },
        {
            name: "key",
            envKey: "AZURE_ACCOUNT_KEY",
            secret: true,
            type: ConfigValueType.STRING
        },
        {
            name: "container",
            envKey: "AZURE_ACCOUNT_CONTAINER",
            secret: false,
            type: ConfigValueType.STRING
        }
    ]
} as Environment;
