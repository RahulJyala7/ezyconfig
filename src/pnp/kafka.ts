import {ConfigValueType} from "../types/config";
import {Environment} from "../types/environment";
import validators from "../validators";

export = {
    name: "kafka",
    properties: [
        {
            name: "clientId",
            envKey: "KAFKA_CLIENT_ID",
            type: ConfigValueType.STRING,
            secret: false,
            validator: validators.isAlphanumeric
        },
        {
            name: "saslMechanism",
            envKey: "KAFKA_SASL_MECHANISM",
            type: ConfigValueType.STRING,
            secret: false
        },
        {
            name: "username",
            envKey: "KAFKA_CLUSTER_API_KEY",
            type: ConfigValueType.STRING,
            secret: false
        },
        {
            name: "password",
            envKey: "KAFKA_CLUSTER_API_SECRET",
            type: ConfigValueType.STRING,
            secret: true
        },
        {
            name: "schemaRegistry",
            envKey: "KAFKA_SCHEMA_REGISTRY_URL",
            type: ConfigValueType.STRING,
            validator: validators.isURL,
            secret: false
        },
        {
            name: "schemaRegistryUser",
            envKey: "KAFKA_SCHEMA_REGISTRY_API_KEY",
            type: ConfigValueType.STRING,
            secret: false
        },
        {
            name: "schemaRegistryPass",
            envKey: "KAFKA_SCHEMA_REGISTRY_API_SECRET",
            type: ConfigValueType.STRING,
            secret: true
        },
        {
            name: "enableSSL",
            envKey: "KAFKA_SSL",
            type: ConfigValueType.BOOLEAN,
            secret: false
        },
        {
            name: "brokers",
            envKey: "KAFKA_BROKERS",
            type: ConfigValueType.STRING,
            secret: false,
            splitOn: ","
        },
        {
            name: "requestTimeout",
            envKey: "KAFKA_REQUEST_TIMEOUT",
            type: ConfigValueType.NUMBER,
            secret: false
        },
        {
            name: "maxRetries",
            envKey: "KAFKA_MAX_RETRIES",
            type: ConfigValueType.NUMBER,
            secret: false
        },
        {
            name: "consumerGroup",
            envKey: "KAFKA_CONSUMER_GROUP",
            type: ConfigValueType.STRING,
            secret: false
        },
        {
            name: "namespace",
            envKey: "KAFKA_NAMESPACE",
            type: ConfigValueType.STRING,
            secret: false
        },
        {
            name: "schemaRegistryClient",
            envKey: "KAFKA_COLLECTION_SERVICE_CLIENT",
            type: ConfigValueType.STRING,
            secret: false
        }
    ]
} as Environment;
