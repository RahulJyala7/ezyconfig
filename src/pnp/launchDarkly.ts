import {ConfigValueType} from "../types/config";
import {Environment} from "../types/environment";

export default {
    name: "launchDarkly",
    properties: [
        {
            name: "apiKey",
            envKey: "LAUNCH_DARKLY_API_KEY",
            secret: false,
            type: ConfigValueType.STRING
        }
    ]
} as Environment;
