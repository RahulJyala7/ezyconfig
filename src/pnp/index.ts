import kafka from "./kafka";
import mongo from "./mongo";
import launchDarkly from "./launchDarkly";
import azure from "./azure";

export = {
    kafka,
    mongo,
    launchDarkly,
    azure
};
