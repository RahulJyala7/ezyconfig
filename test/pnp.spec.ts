import {kafka, azure, launchDarkly, mongo} from "../src/pnp";

describe("Default Plug and Play Environments", () => {
    test("kafka is configured correctly", () => {
        expect(kafka).toMatchSnapshot();
    });

    test("mongo is configured correctly", () => {
        expect(mongo).toMatchSnapshot();
    });

    test("azure is configured correctly", () => {
        expect(azure).toMatchSnapshot();
    });

    test("launch darkly is configured correctly", () => {
        expect(launchDarkly).toMatchSnapshot();
    });
});
