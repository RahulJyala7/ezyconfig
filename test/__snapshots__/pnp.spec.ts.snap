// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default Plug and Play Environments azure is configured correctly 1`] = `
Object {
  "name": "azure",
  "properties": Array [
    Object {
      "envKey": "AZURE_ACCOUNT",
      "name": "account",
      "secret": false,
      "type": "STRING",
    },
    Object {
      "envKey": "AZURE_ACCOUNT_KEY",
      "name": "key",
      "secret": true,
      "type": "STRING",
    },
    Object {
      "envKey": "AZURE_ACCOUNT_CONTAINER",
      "name": "container",
      "secret": false,
      "type": "STRING",
    },
  ],
}
`;

exports[`Default Plug and Play Environments kafka is configured correctly 1`] = `
Object {
  "name": "kafka",
  "properties": Array [
    Object {
      "envKey": "KAFKA_CLIENT_ID",
      "name": "clientId",
      "secret": false,
      "type": "STRING",
      "validator": Object {
        "fn": [Function],
        "name": "isAlphanumeric",
      },
    },
    Object {
      "envKey": "KAFKA_SASL_MECHANISM",
      "name": "saslMechanism",
      "secret": false,
      "type": "STRING",
    },
    Object {
      "envKey": "KAFKA_CLUSTER_API_KEY",
      "name": "username",
      "secret": false,
      "type": "STRING",
    },
    Object {
      "envKey": "KAFKA_CLUSTER_API_SECRET",
      "name": "password",
      "secret": true,
      "type": "STRING",
    },
    Object {
      "envKey": "KAFKA_SCHEMA_REGISTRY_URL",
      "name": "schemaRegistry",
      "secret": false,
      "type": "STRING",
      "validator": Object {
        "fn": [Function],
        "name": "isURL",
      },
    },
    Object {
      "envKey": "KAFKA_SCHEMA_REGISTRY_API_KEY",
      "name": "schemaRegistryUser",
      "secret": false,
      "type": "STRING",
    },
    Object {
      "envKey": "KAFKA_SCHEMA_REGISTRY_API_SECRET",
      "name": "schemaRegistryPass",
      "secret": true,
      "type": "STRING",
    },
    Object {
      "envKey": "KAFKA_SSL",
      "name": "enableSSL",
      "secret": false,
      "type": "BOOLEAN",
    },
    Object {
      "envKey": "KAFKA_BROKERS",
      "name": "brokers",
      "secret": false,
      "splitOn": ",",
      "type": "STRING",
    },
    Object {
      "envKey": "KAFKA_REQUEST_TIMEOUT",
      "name": "requestTimeout",
      "secret": false,
      "type": "NUMBER",
    },
    Object {
      "envKey": "KAFKA_MAX_RETRIES",
      "name": "maxRetries",
      "secret": false,
      "type": "NUMBER",
    },
    Object {
      "envKey": "KAFKA_CONSUMER_GROUP",
      "name": "consumerGroup",
      "secret": false,
      "type": "STRING",
    },
    Object {
      "envKey": "KAFKA_NAMESPACE",
      "name": "namespace",
      "secret": false,
      "type": "STRING",
    },
  ],
}
`;

exports[`Default Plug and Play Environments launch darkly is configured correctly 1`] = `
Object {
  "name": "launchDarkly",
  "properties": Array [
    Object {
      "envKey": "LAUNCH_DARKLY_API_KEY",
      "name": "apiKey",
      "secret": false,
      "type": "STRING",
    },
  ],
}
`;

exports[`Default Plug and Play Environments mongo is configured correctly 1`] = `
Object {
  "name": "mongo",
  "properties": Array [
    Object {
      "envKey": "MONGO_USER",
      "name": "user",
      "secret": false,
      "type": "STRING",
    },
    Object {
      "envKey": "MONGO_PASSWORD",
      "name": "password",
      "secret": true,
      "type": "STRING",
    },
    Object {
      "envKey": "MONGO_CONNECTION_STRING",
      "name": "connection",
      "secret": false,
      "type": "STRING",
    },
    Object {
      "envKey": "MONGO_DATABASE",
      "name": "database",
      "secret": false,
      "type": "STRING",
      "validator": Object {
        "fn": [Function],
        "name": "isAlphanumeric",
      },
    },
  ],
}
`;
